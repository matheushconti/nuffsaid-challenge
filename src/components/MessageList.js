import React, { Component, useState, useEffect } from 'react'
import { Grid, Button, Typography, Divider, Container, makeStyles, withStyles } from '@material-ui/core'
import Api from '../api'
import MessageGrid from './MessageGrid'
import MessageSnackbar from './MessageSnackbar'
import { messageTypes } from '../consts'

const snackbarDefault = {open: false, message: ""}

const StyledButton = withStyles({
  root: {
    backgroundColor: '#88FCA3',
    fontWeight: 'bold',
    "&:hover, &:focus": {
      backgroundColor: '#6ee089'
    },
  }
})(Button)

const useStyles = makeStyles((theme) => ({
  buttonsContent: {
    marginBottom: 40,
    marginTop: 10
  }
}))

export default function MessageList(){
  const classes = useStyles()
  const [isApiStarted, setIsApiStarted] = useState(true)
  const [message, setMessage] = useState({})
  const [messagesError, setMessagesError] = useState([])
  const [messagesWarning, setMessagesWarning] = useState([])
  const [messagesInfo, setMessagesInfo] = useState([])
  const [snackbar, setSnackbar] = useState(snackbarDefault)

  const [api] = useState(new Api({
    messageCallback: (message) => {
      setMessage(message)
    },
  }))

  useEffect(() => {
    api.start()
  },[api])

  useEffect(() => {
    console.log(message)
    switch(message.priority){
      case 1:
        handleOpenSnackbar(message.message)
        setMessagesError([message, ...messagesError])
        return
      case 2:
        setMessagesWarning([message, ...messagesWarning])
        return
      case 3:
        setMessagesInfo([message, ...messagesInfo])
        return
      default:
        return
    }
  },[message])

  const handleStart = () => {
    if (isApiStarted) {
      api.stop()
    } else {
      api.start()
    }
    setIsApiStarted(!isApiStarted)
  }

  const handleClear = () => {
    setMessagesError([])
    setMessagesWarning([])
    setMessagesInfo([])
  }

  
  const handleCloseSnackbar = (event, reason) => {
    if (reason === 'clickaway') {
      return
    }
    setSnackbar(snackbarDefault)
  }
  const handleOpenSnackbar = (message) => {
    setSnackbar(snackbarDefault)
    setSnackbar({open: true, message: message})
  }

  return (
    <Grid container>
      <MessageSnackbar snackbar={snackbar} handleCloseSnackbar={handleCloseSnackbar} />
      
      <Grid item xs={12}>
        <Typography gutterBottom variant="h5">
          nuffsaid.com Coding Challenge
        </Typography>
        <Divider />
      </Grid>
      <Container>
        <Grid container justifyContent="center" className={classes.buttonsContent}>
          <StyledButton
            variant="contained"
            onClick={() => handleStart()}
            style={{marginRight:10}}
          >
            {isApiStarted ? 'Stop' : 'Start'}
          </StyledButton>
          <StyledButton
            variant="contained"
            onClick={() => handleClear()}
          >
            Clear
          </StyledButton>
        </Grid>
        <Grid container direction="row" spacing={2}>
            <MessageGrid messages={messagesError} setMessages={setMessagesError} type={messageTypes.error} />
            <MessageGrid messages={messagesWarning} setMessages={setMessagesWarning} type={messageTypes.warning} />
            <MessageGrid messages={messagesInfo} setMessages={setMessagesInfo} type={messageTypes.info} />
        </Grid>
      </Container>
    </Grid>
  )
}