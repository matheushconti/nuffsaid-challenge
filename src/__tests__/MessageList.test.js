import React from 'react'
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import MessageList from '../components/MessageList'

configure({ adapter: new Adapter() });

describe('Testing MessageList', () => {
    test('rendering MessageList', () => {
        const tree = renderer.create(<MessageList />).toJSON();
        expect(tree).toMatchSnapshot();
    });
})