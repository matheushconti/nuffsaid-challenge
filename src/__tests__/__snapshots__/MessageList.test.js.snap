// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MessageList rendering MessageList 1`] = `
<div
  className="MuiGrid-root MuiGrid-container"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
    >
      nuffsaid.com Coding Challenge
    </h5>
    <hr
      className="MuiDivider-root"
    />
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      className="MuiGrid-root makeStyles-buttonsContent-1 MuiGrid-container MuiGrid-justify-content-xs-center"
    >
      <button
        className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-2 MuiButton-contained"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "marginRight": 10,
          }
        }
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Stop
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-2 MuiButton-contained"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Clear
        </span>
      </button>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      >
        <div
          className="MuiGrid-root makeStyles-titleContainer-3"
        >
          <h5
            className="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
          >
            Error Type 1
          </h5>
          <p
            className="MuiTypography-root MuiTypography-body2"
          >
            Count 0
          </p>
        </div>
        <div
          className="MuiGrid-root"
        />
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      >
        <div
          className="MuiGrid-root makeStyles-titleContainer-3"
        >
          <h5
            className="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
          >
            Warning Type 2
          </h5>
          <p
            className="MuiTypography-root MuiTypography-body2"
          >
            Count 0
          </p>
        </div>
        <div
          className="MuiGrid-root"
        />
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      >
        <div
          className="MuiGrid-root makeStyles-titleContainer-3"
        >
          <h5
            className="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
          >
            Info Type 3
          </h5>
          <p
            className="MuiTypography-root MuiTypography-body2"
          >
            Count 0
          </p>
        </div>
        <div
          className="MuiGrid-root"
        />
      </div>
    </div>
  </div>
</div>
`;
